connections/multiselect2.md

```diff
@@ -22,11 +22,14 @@ Peers advertising a multiaddr that includes a handshake protocol MUST support Mu
 
 #### TCP Simultaneous Open
 
-TCP allows the establishment of a connection if two endpoints start initiating a connection at the same time. This is called TCP Simultaneous Open. For libp2p, this is problematic, since most stream multiplexers assign stream IDs based on the role (client or server) of and endpoint.
+TCP allows the establishment of a connection if two endpoints start initiating a connection at the same time. This is called TCP Simultaneous Open. For libp2p, on the one hand, this poses a problem, since most stream multiplexers assign stream IDs based on the role (client or server) of an endpoint. On the other hand, it can be used as a hole punching technique to facilitate NAT traversal.
 
-It is therefore desirable to fail a connection attempt as early as possible, if a TCP Simultaneous Open occurs. TLS 1.3 and Noise provide this guarantee: For example, the TLS handshake will fail if an endpoint receives a ClientHello instead of a ServerHello as a response to its ClientHello.
+TLS as well as Noise will fail the handshake if both endpoints act as clients. In case of such a handshake failure, the two endpoints need to restart the handshake. Endpoints MUST NOT close the underlying TCP connection in this case. Implementations SHOULD specifically test for this type of handshake failure, and not treat any handshake failure as a potential Simultaneous Open.
 
-Since secio doesn't provide this property, secio cannot be used with Multiselect 2.0.
+To determine the roles in the second handshake attempt, endpoints calculate the SHA-256 hash of the handshake messages that were sent and received (including any error message(s) that the handshake protocol might have sent) during the failed handshake attempt.
```

Comment:
Are these handshake messages available to the application? It seems to me that these will be internal to the TLS implementation and not actually available to hash.

Reply:
You can just wrap the `Read` and `Write` of the `net.Conn`.

Reply:
This is true in go, but what about other languages?
Also, does the "wrap the Read/Write" imply that we hash all the bytes sent/received to make the decision?
It's still now clear how an implementor would implement this part of the spec.

Reply:
These are the bytes sent in the clear over the wire. I'm not familiar with the TLS / Noise implementations in other languages, but I'd be surprised if this posed a major challenge in any language.
@tomaka, @yusefnapora Can you shed some light on this?

Reply:
Any language that uses a TLS library that reads/writes to the file descriptor will have problems with this.

Reply:
We discussed this with @Stebalien and we both think this (hashing the plaintext messages) is the wrong way to approach it.
We totally don't have to do that: we have the peer IDs for both endpoints here, because we need them for the TLS handshake. 
So what we should do is simply pick the peer with the highest (or lowest) numeric peer id to act as the initiator if the handshake fails with unexpected client hello.

Reply:
@vyzo Thank you. Using peer IDs makes a lot of sense to me, and is definitely much simpler than recording messages sent on the wire.
I updated the spec using SHA256(peer ID). I believe that makes numerical comparisons of peer IDs easier, especially if they have different lengths.

Reply:
There's an additional detail we need to specify: When receiving the unexpected handshake message, TLS does two things:
1. It sends a TLS alert to the peer, and
2. it aborts the handshake and returns an error message.

If we just restart the handshake when the handshake message is returned, the peer's alert will still be in flight, and typically will be received after we initiated the second handshake, leading to a handshake failure. There are three ways we can solve this:
1. Modify the TLS stack to not send the alert,
2. Filter out the alert before it reaches the TLS stack
3. Wait for the alert to arrive, and then start the second connection attempt afterwards.

I don't think 1. is very practical, since it requires us to modify the TLS stack. 2. would be doable, but requires wrapping of the connection. 3. costs us half a roundtrip, but could be implemented without modifications to the TLS stack or wrapping of the connection.

Reply:
Update: I just implemented option 3 in https://github.com/libp2p/go-libp2p-tls/compare/tcp-sim-open-option3, and unfortunately, it's inherently racy: When `tls.Conn().Handshake()` returns, we have no idea if this is before or after the TLS alert from the peer was received. We therefore can't reliably catch the TLS alert afterwards, since it might never arrive.

I think there's an option 4:
4. Filter out the `ClientHello` before it arrives at the TLS stack. If we receive one, abort the current handshake (taking care that no abort alert is sent out), and start anew.

Unfortunately, this option requires wrapping of the TLS connection, but at least it should be race-free. Furthermore, it doesn't cost us any additional (half) round-trips.

Reply:
Here's an implementation of option 4, which seems to work: https://github.com/libp2p/go-libp2p-tls/pull/38.
The downside is that we're duplicating a bit of the record layer parsing code of TLS, since we need to make sure that we read the whole `ClientHello` from the connection before we can start the second connection attempt.

Reply:
The concern here is again other language implementations and whether they can do that in reasonable fashion.
Other than that, I am fine with it.

Reply:
I haven't verified this at all, so we shouldn't assume this is correct... but the [docs for node's tls.connect()](https://nodejs.org/api/tls.html#tls_tls_connect_options_callback) show that you can pass in any duplex Stream as the underlying socket which will be used for the connection. 

So it seems like we should be able to wrap a `net.Socket` with some code similar to @marten-seemann's go code which parses the first handshake message and aborts with a custom error if the server sends a `ClientHello`. Then we can do the peer id comparison and retry the connection when we get the special error type.

This won't work in browsers, but I don't think we have a clear path to supporting TLS 1.3 at all there, and we don't have access to raw TCP in browsers anyway, so it's kind of a moot point.